x-pack/legacy/plugins/infra/public/hooks/use_track_metric.tsx

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useEffect } from 'react';
+import { trackUiMetric } from '../../../../../../src/legacy/core_plugins/ui_metric/public';
+
+/**
+ * Note: The UI Metric plugin will take care of sending this data to the telemetry server.
+ * You can find these metrics stored at:
+ * stack_stats.kibana.plugins.ui_metric.{app}.{metric}(__delayed_{n}ms)?
+ * which will be an array of objects each containing a key, representing the metric, and
+ * a value, which will be a counter
+ */
+
+type ObservabilityApp = 'infra_metrics' | 'infra_logs' | 'apm' | 'uptime';
+
+interface TrackProps {
+  app: ObservabilityApp;
+  delay?: number; // in ms
+  effectDependencies?: unknown[];
+}
+
+type TrackMetricProps = TrackProps & { metric: string };
+
+export function useTrackMetric({
+  app,
+  metric,
+  delay = 0,
+  effectDependencies = [],
```

Comment:
While I see that there are probably uses for exposing the naked dependencies like this, I wonder what expectation the consumer of this hook should have about it. Will it be merged with other arguments like `metric` and `app` (it won't as of right now, see below). Will it replace some other current dependencies?

Reply:
My thinking is that if you want those things to control the re-running of the effect, you should explicitly provide them in the `effectDependencies` array, sot hat we leave this choice up to the caller.

Reply:
Perhaps it would be more "hook-like" to have a signature like:
```ts
export function useTrackMetric({ app, metric, delay = 0 }, effectDependencies) {
```
?

Reply:
That would look less surprising to my eyes :+1:

Another thought about the consequence of these dependencies: They cancel and re-start the delay. So if they change more frequently than the delay, it will never trigger. https://overreacted.io/making-setinterval-declarative-with-react-hooks/ shows how to avoid that if we wanted, but we could also do that as a follow-up. (Might be worth factoring out as a hook into `utils` then.)

Reply:
Yeah I thought about this too. I think in most cases I can think of, that's what you want (in the case where you have this hook outside of your routes listening to route changes, if the user didn't stay on the page for `delay` ms before the path changed, then you want to cancel it). 

I talked to @Bamieh and I think we're going to try to get this hook moved into the UI Metric plugin so it's just imported from there the same way you would get access to the function itself.

I'll change this signature to be more hook-like as I suggested!